rfc/src/rfcs/0008-function-contracts.md

```diff
@@ -0,0 +1,580 @@
+input or any `static` global referenced, directly or transitively, by the
+function. While the inferred footprint is sound and enough for successful
+contract checking[^inferred-footprint] it can easily turn large section of
+memory to non-deterministic values, invalidate invariants of your program and
+cause the verification to fail when the contract is used as a stub.
+
+[^inferred-footprint]: While inferred memory footprints are sound for both safe
+    and unsafe rust certain features in unsafe rust (e.g. `RefCell`) get
+    inferred incorrectly and will lead to a failing contract check.
+
+To reduce the scope of havocing Kani provides the `#[kani::assigns(CONDITION,
+ASSIGN_RANGE...)]` and `#[kani::frees(CONDITION, LVALUE...)]` attributes. When
+these attributes are provided Kani will only havoc the location mentioned in
+`ASSIGN_RANGE` and `LVALUE` instead of the inferred footprint. Additionally Kani
+verifies during checking that only the mentioned memory regions are touched and
+only under the specified `CONDITION`. The `CONDITION` is optional and defaults
+to `true`.
+
+`LVALUE` are simple expressions permissible on the left hand side of an
+assignment. They compose of the name of one function argument and zero or more
+projections (dereference `*`, field access `.x`, slice indexing `[1]`).
+
+The `ASSIGN_RANGE` permits any `LVALUE` but additionally permits more complex
+slice expressions as the last projection that applies to pointer values. `[..]`
+denotes the entirety of an allocation, `[i..]`, `[..j]` and `[i..j]` are
+ranges of pointer offsets. A slicing syntax `p[i..j]` only applies if `p` is a
+`*mut T` and points to an array allocation. The slice indices are offsets with
+sizing `T`, e.g. in Rust `p[i..j]` would be equivalent to
+`std::slice::from_raw_parts(p.offset(i), i - j)`. `i` must be smaller or equal
+than `j`.
+
+Because lvalues are restricted to using projections only, Kani must break
+encapsulation here. If need be we can reference fields that are usually hidden,
+without an error from the compiler.
+
+[^side-effects]: Code used in contracts is required to be side effect free which
+    means it must not perform I/O, mutate memory (`&mut` vars and such) or
+    (de)allocate heap memory. This is enforced by the verifier, see the
+    discussion in the [future](#future-possibilities) section.
+
+
+### Detailed Attribute Contraints Overview
+
+Any violation of the following constraints constitutes a compile-time error.
+
+- A function may have any number of `requires`, `ensures`, `assigns` and `frees`
+  attributes. Any function with at least one such annotation is considered as
+  "having a contract".
+
+  Harnesses (general or for contract checking) may not have any such annotation.
+
+- A function may have one `proof_for_contract(TARGET)` annotation. `TARGET` must
+  "have a contract". Multiple functions may have `proof_for_contract`
+  annotations with the same `TARGET`. All such harnesses must pass verification,
```

Comment:
I am not convinced we should enforce this. Maybe we can enforce this for contracts defined in the same crate as the stub_verified is used, but this will be much harder for dependency crates.

I also think this will make `--harness` behavior very awkward, since users won't be able to run their harnesses in isolation.